src/optionals/queue.c

```diff
@@ -0,0 +1,208 @@
+    snprintf(queueString, 8, "<Queue>");
+    return queueString;
+}
+
+
+static Value queueIsFull(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 0) {
+        runtimeError(vm, "size() takes no arguments (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+    bool isFull = queue->capacity == queue->count;
+
+    return BOOL_VAL(isFull);
+}
+
+static Value queueIsEmpty(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 0) {
+        runtimeError(vm, "empty() takes no arguments (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+    bool isFull = queue->count == 0;
+
+    return BOOL_VAL(isFull);
+}
+
+static Value queueCap(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 0) {
+        runtimeError(vm, "cap() takes no arguments (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+
+    return NUMBER_VAL(queue->capacity);
+}
+
+static Value queueLen(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 0) {
+        runtimeError(vm, "len() takes no arguments (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+
+    return NUMBER_VAL(queue->count);
+}
+
+static Value queuePeek(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 0) {
+        runtimeError(vm, "peek() takes no arguments (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+    
+    return newResultSuccess(vm, queue->dq[queue->front]);
+}
+
+static Value queuePush(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 1) {
+        runtimeError(vm, "push() takes 1 argument (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+
+    if (queue->capacity < queue->count + 1) {
+        int oldCapacity = queue->capacity;
+        queue->capacity = GROW_CAPACITY(oldCapacity);
+        queue->dq = GROW_ARRAY(vm, queue->dq, Value, oldCapacity, queue->capacity);
+    }
+
+    if(queue->rear == queue->capacity-1) {
+        queue->rear = -1;            
+    }
+
+    queue->dq[++queue->rear] = args[1];
+    queue->count++;
+
+    return newResultSuccess(vm, NIL_VAL);
+}
+
+static Value queuePop(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 0) {
+        runtimeError(vm, "pop() takes no arguments (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+
+    Value data = queue->dq[queue->front++];
+
+    if (queue->front == queue->capacity) {
+        queue->front = 0;
+    }
+
+    queue->count--;
+
+    if (queue->count < queue->capacity) {
+        int oldCapacity = queue->capacity;
+        queue->capacity = SHRINK_CAPACITY(oldCapacity);
+        queue->dq = SHRINK_ARRAY(vm, queue->dq, Value, oldCapacity, queue->capacity);
+    }
+
+    return newResultSuccess(vm, data);
```

Comment:
```suggestion

    return data;
```

We should definitely think about sizing down queue sizes if it meets a certain threshold, but we wouldn't want it to happen each call to `pop()` (as count is always going to be less than capacity in this case)

Something to think about with lists too actually, I don't think they size down

Reply:
Curious your thoughts on this. I changed the shrink operation to only shrink the underlying array if the count is less than half the capacity but greater than the default size.

Reply:
I like it yep!